---
title: "Project 2D Lydia Chen"
output:
  html_document:
    df_print: paged
---

# Introduction
## Scientific Question
Are the amino acid sequences and thus the structure of the DNA binding domain of the human FOXO3 gene more than 70% identical to the FOXO3 gene in mice, pigs, and chimpanzees?

## Background
FOXO3 is a gene that is a part of the forkhead family of transcription factors. Specifically, FOXO3 encodes the transcription factor forkhead box O-3 (Morris et al., 2015). It has long been associated with longevity in humans, and many studies have been done on its relation to healthy aging. Forkhead box proteins are involved in many different cellular functions, such as cellular proliferation, and they have domains that are specific for DNA binding (Myatt et al., 2007). Notably, the "O" class forkhead proteins, of which FOXO3 is a part of, all are highly conserved, and they all have a forkhead DNA-binding domain (Tsai et al., 2007). FOXO3 specifically is involved in the PI3K/Akt signaling pathway. When it is not phosphorylated, FOXO3 localizes at the nucleus to regulate gene transcription (Stefanetti et al., 2018). However, when the PI3K//Akt pathway phosphorylates FOXO3, it then leaves the nucleus and becomes inactive (Lu et al., 2018). 

Due to the many roles of FOXO3 and its importance in longevity and healthy aging in humans, as well as its involvement in transcriptional activity, it is interesting to see whether the DNA binding domain of the human FOXO3 gene is highly conserved in other species as well. The data for the multiple sequence alignment is from Uniprot, while the data for homology modeling is from PDB (accession number 2UZK). 

## Scientific Hypothesis
If FOXO3's DNA binding domain is highly conserved across humans, mice, pigs, and chimpanzees, then we expect the human FOXO3 protein sequence to be more than 70% similar to FOXO3 in other species, and the structure of the DNA-binding domain to not be affected by differences in sequence.

## Description
To look at the similarity of the FOXO3 amino acid sequence between different species, I performed multiple sequence alignment on the amino acid sequences of the human, mouse, pig, and chimpanzee FOXO3 gene. To download the data for multiple sequence alignment, I went to Uniprot and searched the database for "FOXO3", and downloaded the 4 sequences into one fasta file. The entry numbers for Uniprot are:
Human: O43524
Mouse: Q9WVH4
Pig: F1RT23
Chimpanzee: H2R710
To visualize the similarity of the sequences, I analyzed the data from multiple sequence alignment using sequence logos, and focused on the section between amino acids 157 to 252, since this is where the DNA binding domain is located. 

Next, to look at the structures of the DNA binding domains for FOXO3 in each of the species, I used homology modeling to model the DNA binding domain of FOXO3 in mice, pigs, and chimpanzees on top of the existing PDB structure for human FOXO3. I did this by using SWISS-MODEL and inputting the FASTA amino acid sequences of the other three species to predict their models using human FOXO3 as a template. I then used NGLVieweR to individually view all four of the structures: the human FOXO3 DNA binding domain (PDB accession: 2UZK), and the three modeled structures of the other species. I then used PyMol to view all four structures together, so that they are shown on top of one another to see if any differences are present. I then used 3D protein measurements to measure the length of the DNA binding domain by finding the distance between the first residue, Gly157, and the last residue, Ser252, of the DNA binding domain. By doing so, I can see whether or not the DNA binding domains are of the same size structurally.

# Loading in Packages

```{r}
# 1. The package "Biostrings" from Bioconductor is required for the function readAAStringSet(), 
# and it is used for the manipulation of biological sequences.
# if (!require("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")

#BiocManager::install("Biostrings")
library(Biostrings)

# 2. The package "msa" from Bioconductor is needed to run multiple sequence alignment 
# using the function msa(). It will align two or more amino acid/DNA sequences in a fasta file. 
# It is also needed to run msaPrettyPrint(), which will print out a color-coded version of the
# multiple sequence alignment and the sequence logos.
# if (!require("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")

#BiocManager::install("msa")
library(msa)

# 3. The package "bio3d" is needed to perform 3D protein measurements from the structures.
# Specifically, it is needed to read in the pdb files using read.pdb() and to calculate the 3D distance
# between two parts/atoms/residues of the structure using dist.xyz().
# install.packages("bio3d", dependencies=TRUE)
library(bio3d)

# 4. The package "seqinr" is needed to read in fasta files, so that the sequences of genes can be analyzed.
# install.packages('seqinr')
library(seqinr)

# 5. The package "remotes" makes it possible to install packages that are stored in remote repositories,
# such as Github, which is where NGLVieweR is available. The function remotes::install_github("nvelden/NGLVieweR")
# will then install NGLVieweR, which is needed to visualize the structure of the human FOXO3 gene's DNA binding
# domain by using the function NGLVieweR(). 
# install.packages("remotes")
# remotes::install_github("nvelden/NGLVieweR")
library(NGLVieweR)
```

# Bioinformatics Analysis - Multiple Sequence Alignment
Multiple sequence alignment refers to the alignment of three or more DNA or amino acid sequences. It uses algorithms in order to identify which amino acids/nucleotides are the same between the different sequences, and where there are indels or mutations in the sequence. Mutiple sequence alignment can be used to identify evolutionary relationships between different genes, or to see where the conserved regions are between the gene in different species, which is what I am doing here. To run multiple sequence alignment in R using the msa() function, you need to read in a fasta file containing the fasta sequences of each sequence you want to align together. You then want to convert this to an object of class XStringSet to use for the msa() function, and since I am running amino acid sequences, I want to convert it to an AAStringSet. 

```{r}
# Set up the global variable "FOXO3_unaligned", which is an AAStringSet variable consisting of the amino acid sequences of the FOXO3 gene in different species. This is a global variable because it will be used throughout the code so that in the future, we do not have to type out "readAAStringSet("/Users/lydiachen/Downloads/FOXO3_species.fasta")" and instead can just call the variable "FOXO3_unaligned" that we defined.

# The function readAAStringSet() is a built-in function from Bioconductor Biostrings that reads FASTA files in an XStringSet object. In this case, my fasta sequences are amino acid sequences, so I would apply AAStringSet. This acknowledges that my fasta file contains amino acid sequences, and this can then be used in multiple sequence alignment.
FOXO3_unaligned <- readAAStringSet("/Users/lydiachen/Downloads/FOXO3_species.fasta")

# check datatype
typeof(FOXO3_unaligned)

# identify how many sequences there are in the fasta file, thus how many sequences we are using in the msa.
length(FOXO3_unaligned)

# run the multiple sequence alignment using the msa() function, as a part of the msa package in Bioconductor
# define my own function to satisfy "function written that performs some part of the bioinformatics analyses". This function, which I name "seqaln", takes in a variable called "unaligned" and then performs msa on it using the function msa(). The result of this is then named as the variable "msalign", and return(msalign) means that the output will be the result of the multiple sequence alignment. In this function, "unaligned" is a local variable, since it is only defined within the function seqaln(). Outside of the function, "unaligned" will not be assigned any value, so it is nota global variable. We can also double check this by seeing that it does not appear in the Global Environment after we run the code. 
seqaln <- function(unaligned) {
  msalign <- msa(unaligned)
  return(msalign)
}

# use the function defined above seqaln() on our fasta file FOXO3_unaligned, so that multiple sequence alignment is run on the four amino acid sequences in the fasta file. Name the result as "FOXO3_msa", which is a global variable since it is now defined throughout the notebook as the result of seqaln(FOXO3_unaligned), and also appears in the global environment after running the code.
FOXO3_msa <- seqaln(FOXO3_unaligned)

# print the results of the multiple sequence alignment. The order is human, chimpanzee, pig, and mouse.
print(FOXO3_msa, show = "complete")
```

# Data Analysis - Sequence Logos
Sequence logos are used to visualize sequence conservation from multiple sequence alignment. After msa is performed, we will know which regions of the sequence are most conserved (i.e. the sequence is identical/very similar between sequences), and this can be shown in a sequence logo. The sequence logo will take the consensus sequence, which is the sequence that contains the most frequently seen residue in each spot from the alignment, and show the diversity of the sequences. The bigger a letter is in a specific position, the more frequently it appears in all the sequences at that location. To create a sequence logo, we will need to input the data from the multiple sequence alignment in order to see which residues are most frequent in each spot, and which areas are the most conserved. This information will then be used to create the sequence logo.

```{r}
# The function msaPrettyPrint creates sequence logos from msa data. I run msaPrettyPrint using FOXO3_msa, which is a global variable containing the multiple sequence alignment results, and the results from msaPrettyPrint will be generated as a "tex" file. This will create a LaTex file that will include the sequence logo. To actually visualize the sequence logo, I set showLogo = "top" to display the letters on top of the msa, and setting verbose = TRUE will print the progress messages. 

# Source: https://bioconductor.org/packages/devel/bioc/vignettes/msa/inst/doc/msa.pdf
# Note: I originally set output = "pdf", which was able to run and produce a pdf file with the sequence logo, but for some reason, it would not knit. Therefore, I first ran the code as output = "pdf" to get the pdf file, and then took a screenshot of the PDF file and embedded it in this R notebook below, since I was unable to embed a PDF file in the notebook and actually view it in the knit HTML/PDF.
msaPrettyPrint(FOXO3_msa, output="tex", showNames="none", showLogo="top",
consensusColor="ColdHot", showLegend=FALSE,
askForOverwrite=FALSE, verbose = TRUE)

```

! [sequence logo](images/FOXO3msa.png)

# Bioinformatics Analysis - Homology Modeling
Homology modeling refers to using an amino acid sequence for which the structure is unknown and a template structure, usually a related protein, to predict the 3D structure of the amino acid sequence. There are multiple steps involved in this process. First, sequence alignment is needed to identify the conserved regions and differences in the sequences of the template and the target, Then, these results are used to create the backbone of the new structure, and further modeling of the loops and side chains is performed. Lastly, the model is optimized and validated. To perform homology modeling, I first searched the PDB database for the known structure of the human FOXO3 gene's DNA binding domain. I then used SWISS-MODEL to predict the structures of FOXO3 for the other three species, since their structures are unknown, and the human FOXO3 gene's structure was used as the template. To use SWISS-MODEL, the amino acid sequences in fasta format are inputted into SWISS-MODEL, and SWISS-MODEL searches the PDB database for a known template. Once SWISS-MODEL produces the predicted structures, I saved them as PDB files. I used these PDB files to view the four structures (one for each of the 4 species) separately using NGLVieweR, and then I opened all four PDB files in the same PyMol window to view the four structures modeled on top of each other, to see if there are any differences in structure. 

```{r}
# View FOXO3 DNA-binding domains of human, mouse, pig, and chimpanzee using NGLVieweR. Load human protein by PDB code for function NGLVieweR(), or by PDB files for the predicted structures for other species.
# To visualize it in the cartoon mode, use the function addRepresentation("cartoon").

# Note: here I tried to run a for loop for the three modeled structures, but when I run the code and knit the notebook to HTML, the output of the for loop is just a black screen (the background of the NGLVieweR), and the structures cannot be viewed.
for (read in c("mouse_FOXO3.pdb", "pig_FOXO3.pdb", "chimpanzee_FOXO3.pdb")) {
  print(NGLVieweR(read, format = ".pdb") %>%
     addRepresentation("cartoon"))
}

# view structure of human FOXO3 DNA binding domain
NGLVieweR("2UZK") %>%
addRepresentation("cartoon")

# view structure of mouse FOXO3 DNA binding domain
NGLVieweR("mouse_FOXO3.pdb") %>%
addRepresentation("cartoon")

# view structure of pig FOXO3 DNA binding domain
NGLVieweR("pig_FOXO3.pdb") %>%
addRepresentation("cartoon")

# view structure of chimpanzee FOXO3 DNA binding domain
NGLVieweR("chimpanzee_FOXO3.pdb") %>%
addRepresentation("cartoon")
```

Next, view all four structures together in PyMol. I first used "fetch 2UZK" in PyMol to view the structure from the PDB database of the human FOXO3 DNA binding domain, and since in the database the structure includes the DNA binding domain sequence twice, as well as the DNA it is bound to, I deleted the DNA binding domain sequence repeat and the DNA by first typing "set seq_view, 1" into the console to display the sequence, and then selecting the residues and selecting "delete atoms". I then opened the other three PDB files in the same PyMol window, and I colored each species a different color: human = red, mouse = green, pig = blue, and chimpanzee = magenta in order to clearly see the structures of each. However, since the structures are completely the same, due to them having the same sequence, I was only able to see the red and green (human and mouse) colors in the structures. This may be because I first fetched the human structure and then added the mouse structure, so these are the two that are displayed since I added them first. I have attached an exported image of the four modeled structures together in PyMol and embedded them here.

! [PyMol screenshot of the 4 modeled structures](images/pymol_four.png)

# Data Analysis - 3D Protein Measurements
3D protein measurements are used to further analyze the 3D structure of a protein, especially if it is a predicted model. It can be used to measure the distance between certain residues or domains, and by doing so, we can compare the structures of the same region in different structures. In this case, we want to compare the DNA binding domains of FOXO3 in different species, namely humans, mice, pigs, and chimpanzees. To do so, one 3D protein measurement we can do is measure the distance between the first residue of the DNA binding domain, Gly157, and the last residue of the DNA binding domain, Ser252. 

```{r}
# use for loop to avoid redundant code. "2uzk.pdb" is the file for human, which is the PDB file that is directly downloaded from the PDB database. This for loop states that for each file, read in the file using read.pdb().
# The variable "seq" is a global variable, since it appears in the global environment when you run the code. It refers to the four pdb files for each of the species when they are read using the function read.pdb(). Similarly, the variables "coord1", "coord2", and measurements are all global variables as well.

for (file in c("2uzk.pdb", "mouse_FOXO3.pdb", "pig_FOXO3.pdb", "chimpanzee_FOXO3.pdb")) {
  # read in each pdb file using read.pdb()
  seq <- read.pdb(file)

  # set the coordinates of Gly157 and Ser252, the beginning and end of the DNA binding domain, which is their location in 3D (x,y,z).
  coord1 <- seq$atom[157, c("x","y","z")]
  coord2 <- seq$atom[252, c("x","y","z")]

  # calculate the distance between the two 3D coordinates to determine how far apart they are, and print this distance.
  measurements <- dist.xyz(coord2, coord1)
  print(measurements)
}
```

# Analyzing the Results
First off, looking at the multiple sequence alignment, it appears that the sequences for human, mouse, pig, and chimpanzee are highly conserved throughout the entire FOXO3 gene. There are pretty much no differences in the sequences, except for a couple spots where there are differing residues. However, there are no differing residues at all within the DNA binding domain, which is between residues 157 and 252, which means that all four species have the exact same sequence in the DNA binding domain. Similarly, from the homology modeling, we can see that the 3D measurements for the DNA binding domain of all four species is the same, except for the human FOXO3 gene, which I have tried to troubleshoot, but for some reason is still different from the others. I believe that this may be due to the PDB file itself and some differences in formatting between a PDB file directly downloaded from the PDB database and those produced from SWISS-MODEL. However, since the sequences in the DNA binding domains for the four species are exactly the same, we can assume that the structures should also be exactly the same. This means that the DNA binding domain for all four species is the same size from its first to last residue, which makes sense since the amino acid sequence is exactly the same. Thus, to answer my scientific question and address my scientific hypothesis, the amino acid sequences and the structure of the DNA binding domain of the human FOXO3 gene are definitely more than 70% identical to the FOXO3 gene in mice, pigs, and chimpanzees. In fact, they are more than likely identical, meaning that the FOXO3 gene is highly conserved in all these species.  
